### PR TITLE
`azurerm_windows_web_app`, `azurerm_windows_web_app_slot`, `azurerm_linux_web_app`, `azurerm_linux _web_app_slot` - fix auto heal slow request with path issue

### PR DESCRIPTION
**Pull request by xiaxyi**
In current provider, we only allow user to set the slow request rule with one path, but slow request rule with multiple paths is also supported by api, checking the property - [slowRequestsWithPath](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/web/resource-manager/Microsoft.Web/stable/2021-02-01/CommonDefinitions.json#L437):

The previous PR #18227 was closed but the issue was not fixed.

fix https://github.com/hashicorp/terraform-provider-azurerm/issues/19256
fix https://github.com/hashicorp/terraform-provider-azurerm/issues/22492

Local Acc test:
```
--- PASS: TestAccLinuxWebAppSlot_withAutoHealSlowRequestWithPath (428.62s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/appservice    429.724s
--- PASS: TestAccWindowsWebAppSlot_withAutoHealSlowRequestWithPath (420.55s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/appservice    421.585s
--- PASS: TestAccLinuxWebApp_withAutoHealRulesSlowRequestWithPath (399.84s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/appservice    400.520s
--- PASS: TestAccWindowsWebApp_withAutoHealRulesSlowRequestWithPath (403.73s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/appservice    404.528s
```

image

### CODE REVIEW

**Comment by pietersap** [comment]
I can confirm that the linked issue 17862 is not fixed yet, even though the previous PR was closed because it was supposedly superseded by another one.

Tried today in Azure provider 3.39.1, the autoheal settings (sub_status and win32_status) are not set, and behaviour is still the same as described in the original issue.

**Review by jackofallops — CHANGES_REQUESTED** [review]
Hi @xiaxyi - Thanks for this and apologies again for the delayed review.

I think we have a design issue here. The two properties are present in the swagger, and as far as I can tell, both should be maintained / supported? Which means we should be able to simply add this as a new property? possibly as `ConflictsWith` on `slow_requests` if appropriate to only have one? WDYT?

**Comment by xiaxyi** [comment]
Thanks @jackofallops for the suggestion, I didn't remove the `slow_request` property and I did added the `slow_request_with_path` as a new property. Not sure what do you mean by "add it as a new property"?

**Comment by pietersap** [comment]
Looking forward to this merge

File: website/docs/r/windows_web_app.html.markdown [hunk 1]
```diff
@@ -512,6 +512,20 @@ A `slow_request` block supports the following:
 
 * `path` - (Optional) The path for which this slow request rule applies.
```
Comment (jackofallops) [inline website/docs/r/windows_web_app.html.markdown:513]: We should remove this from the docs as is it being deprecated
```suggestion
```

**Review by jackofallops — CHANGES_REQUESTED** [review]
Hi @xiaxyi - Thanks for this. I've left some comments below if you can take a look?

Thanks!

File: internal/services/appservice/helpers/auto_heal.go [hunk 2]
```diff
@@ -44,6 +46,13 @@ type AutoHealSlowRequest struct {
 	Path      string `tfschema:"path"`
 }
 
+type AutoHealSlowRequestWithPath struct {
+	TimeTaken string `tfschema:"time_taken"`
+	Interval  string `tfschema:"interval"`
+	Count     int    `tfschema:"count"`
+	Path      string `tfschema:"path"`
+}
```
[the remark this answers is not in the document]
Comment (xiaxyi) [reply internal/services/appservice/helpers/auto_heal.go:54]: Thanks @jackofallops for the comment, I removed the path in 4.0 provider and the schema in 4.0 provider is defined as:
```
				"slow_request": {
					Type:     pluginsdk.TypeList,
					Computed: true,
					Elem: &pluginsdk.Resource{
						Schema: map[string]*pluginsdk.Schema{
							"time_taken": {
								Type:     pluginsdk.TypeString,
								Computed: true,
							},

							"interval": {
								Type:     pluginsdk.TypeString,
								Computed: true,
							},

							"count": {
								Type:     pluginsdk.TypeInt,
								Computed: true,
							},
						},
					},
				},

				"slow_request_with_path": {
					Type:     pluginsdk.TypeList,
					Computed: true,
					Elem: &pluginsdk.Resource{
						Schema: map[string]*pluginsdk.Schema{
							"time_taken": {
								Type:     pluginsdk.TypeString,
								Computed: true,
							},

							"interval": {
								Type:     pluginsdk.TypeString,
								Computed: true,
							},

							"count": {
								Type:     pluginsdk.TypeInt,
								Computed: true,
							},

							"path": {
								Type:     pluginsdk.TypeString,
								Computed: true,
							},
						},
					},
				},
			},
		},
	}

```

File: internal/services/appservice/helpers/auto_heal.go [hunk 3]
```diff
@@ -274,6 +283,32 @@ func autoHealTriggerSchemaWindows() *pluginsdk.Schema {
...
+
+				"slow_request_with_path": {
+					Type:     pluginsdk.TypeList,
+					Optional: true,
+					Elem: &pluginsdk.Resource{
+						Schema: map[string]*pluginsdk.Schema{
+							"time_taken": {
+								Type:     pluginsdk.TypeString,
+								Required: true,
+								// ValidateFunc: validation.IsRFC3339Time,
```
[the remark this answers is not in the document]
Comment (xiaxyi) [reply internal/services/appservice/helpers/auto_heal.go:298]: done

File: internal/services/appservice/helpers/auto_heal.go [hunk 4]
```diff
@@ -274,6 +283,32 @@ func autoHealTriggerSchemaWindows() *pluginsdk.Schema {
...
+							},
+
+							"interval": {
+								Type:     pluginsdk.TypeString,
+								Required: true,
+								// ValidateFunc: validation.IsRFC3339Time,
```
[the remark this answers is not in the document]
Comment (xiaxyi) [reply internal/services/appservice/helpers/auto_heal.go:304]: done

File: internal/services/appservice/helpers/auto_heal.go [hunk 5]
```diff
@@ -286,10 +321,45 @@ func autoHealTriggerSchemaWindows() *pluginsdk.Schema {
...
+		s.Elem.(*pluginsdk.Resource).Schema["slow_request"] = &pluginsdk.Schema{
+			Type:     pluginsdk.TypeList,
+			Optional: true,
+			MaxItems: 1,
+			Elem: &pluginsdk.Resource{
+				Schema: map[string]*pluginsdk.Schema{
+					"time_taken": {
+						Type:     pluginsdk.TypeString,
+						Required: true,
+						// ValidateFunc: validation.IsRFC3339Time,
```
[the remark this answers is not in the document]
Comment (xiaxyi) [reply internal/services/appservice/helpers/auto_heal.go:334]: done

File: internal/services/appservice/helpers/auto_heal.go [hunk 6]
```diff
@@ -286,10 +321,45 @@ func autoHealTriggerSchemaWindows() *pluginsdk.Schema {
...
+					},
+
+					"interval": {
+						Type:     pluginsdk.TypeString,
+						Required: true,
+						// ValidateFunc: validation.IsRFC3339Time,
```
[the remark this answers is not in the document]
Comment (xiaxyi) [reply internal/services/appservice/helpers/auto_heal.go:340]: done

File: internal/services/appservice/helpers/auto_heal.go [hunk 7]
```diff
@@ -286,10 +321,45 @@ func autoHealTriggerSchemaWindows() *pluginsdk.Schema {
...
+					},
+
+					"count": {
+						Type:         pluginsdk.TypeInt,
+						Required:     true,
+						ValidateFunc: validation.IntAtLeast(1),
+					},
+
+					"path": {
```
[the remark this answers is not in the document]
Comment (xiaxyi) [reply internal/services/appservice/helpers/auto_heal.go:349]: done

File: internal/services/appservice/helpers/linux_web_app_schema.go [hunk 8]
```diff
@@ -615,6 +617,32 @@ func autoHealTriggerSchemaLinux() *pluginsdk.Schema {
...
+
+				"slow_request_with_path": {
+					Type:     pluginsdk.TypeList,
+					Optional: true,
+					Elem: &pluginsdk.Resource{
+						Schema: map[string]*pluginsdk.Schema{
+							"time_taken": {
+								Type:     pluginsdk.TypeString,
+								Required: true,
+								// ValidateFunc: validation.IsRFC3339Time,
```
[the remark this answers is not in the document]
Comment (xiaxyi) [reply internal/services/appservice/helpers/linux_web_app_schema.go:632]: done

File: internal/services/appservice/helpers/linux_web_app_schema.go [hunk 9]
```diff
@@ -615,6 +617,32 @@ func autoHealTriggerSchemaLinux() *pluginsdk.Schema {
...
+							},
+
+							"interval": {
+								Type:     pluginsdk.TypeString,
+								Required: true,
+								// ValidateFunc: validation.IsRFC3339Time,
```
[the remark this answers is not in the document]
Comment (xiaxyi) [reply internal/services/appservice/helpers/linux_web_app_schema.go:638]: done

File: internal/services/appservice/helpers/linux_web_app_schema.go [hunk 10]
```diff
@@ -627,10 +655,45 @@ func autoHealTriggerSchemaLinux() *pluginsdk.Schema {
...
+		s.Elem.(*pluginsdk.Resource).Schema["slow_request"] = &pluginsdk.Schema{
+			Type:     pluginsdk.TypeList,
+			Optional: true,
+			MaxItems: 1,
+			Elem: &pluginsdk.Resource{
+				Schema: map[string]*pluginsdk.Schema{
+					"time_taken": {
+						Type:     pluginsdk.TypeString,
+						Required: true,
+						// ValidateFunc: validation.IsRFC3339Time,
```
[the remark this answers is not in the document]
Comment (xiaxyi) [reply internal/services/appservice/helpers/linux_web_app_schema.go:668]: done

File: internal/services/appservice/helpers/linux_web_app_schema.go [hunk 11]
```diff
@@ -627,10 +655,45 @@ func autoHealTriggerSchemaLinux() *pluginsdk.Schema {
...
+					},
+
+					"interval": {
+						Type:     pluginsdk.TypeString,
+						Required: true,
+						// ValidateFunc: validation.IsRFC3339Time,
```
[the remark this answers is not in the document]
Comment (xiaxyi) [reply internal/services/appservice/helpers/linux_web_app_schema.go:674]: done

File: internal/services/appservice/linux_web_app_resource_test.go [hunk 12]
```diff
@@ -2734,6 +2749,50 @@ resource "azurerm_linux_web_app" "test" {
...
+  service_plan_id     = azurerm_service_plan.test.id
+  site_config {
+    auto_heal_enabled = true
+    auto_heal_setting {
+      trigger {
+        slow_request {
+          count      = "10"
+          interval   = "00:10:00"
+          time_taken = "00:00:10"
+          path       = null
```
[the remark this answers is not in the document]
Comment (xiaxyi) [reply internal/services/appservice/linux_web_app_resource_test.go:2771]: removed

File: internal/services/appservice/linux_web_app_slot_resource_test.go [hunk 13]
```diff
@@ -1124,6 +1138,49 @@ resource "azurerm_linux_web_app_slot" "test" {
...
+
+  site_config {
+    auto_heal_enabled = true
+    auto_heal_setting {
+      trigger {
+        slow_request {
+          count      = "10"
+          interval   = "00:10:00"
+          time_taken = "00:00:10"
+          path       = null
```
[the remark this answers is not in the document]
Comment (xiaxyi) [reply internal/services/appservice/linux_web_app_slot_resource_test.go:1159]: removed

File: internal/services/appservice/windows_web_app_resource_test.go [hunk 14]
```diff
@@ -2606,6 +2621,50 @@ resource "azurerm_windows_web_app" "test" {
...
+  service_plan_id     = azurerm_service_plan.test.id
+  site_config {
+    auto_heal_enabled = true
+    auto_heal_setting {
+      trigger {
+        slow_request {
+          count      = "10"
+          interval   = "00:10:00"
+          time_taken = "00:00:10"
+          path       = null
```
[the remark this answers is not in the document]
Comment (xiaxyi) [reply internal/services/appservice/windows_web_app_resource_test.go:2643]: removed

File: internal/services/appservice/windows_web_app_slot_resource_test.go [hunk 15]
```diff
@@ -964,6 +978,49 @@ resource "azurerm_windows_web_app_slot" "test" {
...
+
+  site_config {
+    auto_heal_enabled = true
+    auto_heal_setting {
+      trigger {
+        slow_request {
+          count      = "10"
+          interval   = "00:10:00"
+          time_taken = "00:00:10"
+          path       = null
```
[the remark this answers is not in the document]
Comment (xiaxyi) [reply internal/services/appservice/windows_web_app_slot_resource_test.go:999]: removed

File: website/docs/r/windows_web_app.html.markdown (hunk 1 above)
Comment (xiaxyi) [reply website/docs/r/windows_web_app.html.markdown:513]: I added the deprecation message right below this property, shall we still remove it from the docs?

**Comment by xiaxyi** [comment]
Thanks @jackofallops for the review, I updated the code, added the validation and deprecation message. 

Regarding the documentation, shall we still remove the `path` field even though I added the deprecation message right below the property?

**Comment by jackofallops** [comment]
Hi @xiaxyi - Apologies for the delay in re-review. I have found there's a crash behind the feature flag here, which I don't believe can be addressed in this PR as the schema and the underlying go structs will no longer be compatible when the feature flag is set to true, so I'm going to mark this as `blocked` for now whilst the team investigates how to mitigate this.

**Comment by xiaxyi** [comment]
@jackofallops I don't see any crash when running the tests, what about from your side? I see Tom removed the Block tag, would you like to share some updates with me? Thanks!

**Comment by katbyte** [comment]
Now that the API has been updated and resources migrated can you please resolve the conflicts here?

File: website/docs/r/linux_web_app_slot.html.markdown [hunk 16]
```diff
@@ -790,7 +804,9 @@ A `trigger` block supports the following:
 
 * `requests` - (Optional) A `requests` block as defined above.
 
-* `slow_request` - (Optional) One or more `slow_request` blocks as defined above.
+* `slow_request` - (Optional) A `slow_request` blocks as defined above.
```
Comment (jackofallops) [inline website/docs/r/linux_web_app_slot.html.markdown:807]:
```suggestion
* `slow_request` - (Optional) A `slow_request` block as defined above.
```

File: website/docs/r/windows_web_app.html.markdown [hunk 17]
```diff
@@ -773,6 +773,20 @@ A `slow_request` block supports the following:
...
 
+~> **NOTE:** `path` in `slow_request` block will be deprecated in 4.0 provider. Please use `slow_request_with_path` to set a slow request trigger with path specified.
```
Comment (jackofallops) [inline website/docs/r/windows_web_app.html.markdown:776]:
```suggestion
~> **NOTE:** The `path` property in the `slow_request` block is deprecated and will be removed in 4.0 of provider. Please use `slow_request_with_path` to set a slow request trigger with path specified.
```

File: website/docs/r/windows_web_app.html.markdown [hunk 18]
```diff
@@ -821,7 +835,9 @@ A `trigger` block supports the following:
 
 * `requests` - (Optional) A `requests` block as defined above.
 
-* `slow_request` - (Optional) One or more `slow_request` blocks as defined above.
+* `slow_request` - (Optional) A `slow_request` blocks as defined above.
```
Comment (jackofallops) [inline website/docs/r/windows_web_app.html.markdown:838]:
```suggestion
* `slow_request` - (Optional) A `slow_request` block as defined above.
```

File: website/docs/r/windows_web_app_slot.html.markdown [hunk 19]
```diff
@@ -770,6 +770,20 @@ A `slow_request` block supports the following:
 
 * `path` - (Optional) The path for which this slow request rule applies.
 
+~> **NOTE:** `path` in `slow_request` block will be deprecated in 4.0 provider. Please use `slow_request_with_path` to set a slow request trigger with path specified.
```
Comment (jackofallops) [inline website/docs/r/windows_web_app_slot.html.markdown:773]:
```suggestion
~> **NOTE:** The `path` property in the `slow_request` block is deprecated and will be removed in 4.0 of provider. Please use `slow_request_with_path` to set a slow request trigger with path specified.
```

File: website/docs/r/windows_web_app_slot.html.markdown [hunk 20]
```diff
@@ -810,7 +824,9 @@ A `trigger` block supports the following:
 
 * `requests` - (Optional) A `requests` block as defined above.
 
-* `slow_request` - (Optional) One or more `slow_request` blocks as defined above.
+* `slow_request` - (Optional) A `slow_request` blocks as defined above.
```
Comment (jackofallops) [inline website/docs/r/windows_web_app_slot.html.markdown:827]:
```suggestion
* `slow_request` - (Optional) A `slow_request` block as defined above.
```

**Review by jackofallops — CHANGES_REQUESTED** [review]
Hi @xiaxyi - Thanks for this PR, and sorry it fell between the cracks for so long. There's some changes to take a look at below, but if you can get those sorted, I think this will be good to go.

Thanks!

**Comment by xiaxyi** [comment]
@jackofallops Thanks for the review, I updated the code as comments. Let me know if you have any other feedbacks.

**Review by jackofallops — APPROVED** [review]
Thanks @xiaxyi - Apologies for the _very_ long tail on this one. LGTM now 👍

**Comment by jackofallops** [comment]
Tests looking fine, failures are transient errors:

image

**Pull request merged**